### PR TITLE
[srds] Skip unchanged scopes in SRDS config update

### PR DESCRIPTION
**Pull request by stevenzzzz**
Signed-off-by: Xin Zhuang <stevenzzz@google.com>

Commit Message: Skip unchanged scopes in SRDS config update
Additional Description: Also lower the log level when rds_impl ignores a deletion from config server to trace. 
Risk Level: LOW 
Testing: unit test.
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue] #19908
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/common/router/scoped_rds.cc
Comment (chaoqin-li1123) [inline source/common/router/scoped_rds.cc:289]: Can we compare the config directly here? It is possible that two different config has the same name and hash.
(scope_info_iter->second->configProto() == scoped_route_config) should work.

Comment (chaoqin-li1123) [reply source/common/router/scoped_rds.cc:289]: https://github.com/envoyproxy/envoy/blob/a9c3a9663539097d7e1ec67036e2306cb553bdc5/source/common/protobuf/utility.h#L228

**Review by stevenzzzz — COMMENTED** [review]
thanks!

**Review by yanavlasov — DISMISSED** [review]

**Review by chaoqin-li1123 — APPROVED** [review]

File: source/common/router/scoped_rds.cc
Comment (adisuissa) [reply source/common/router/scoped_rds.cc:289]: Relevant issue: #6252
FWIW, I think the main reason `Equivalent` is not typically used is because the config proto isn't being kept.

File: test/common/router/scoped_config_impl_test.cc
Comment (adisuissa) [inline test/common/router/scoped_config_impl_test.cc:394]: nit: const

File: source/common/rds/rds_route_config_subscription.cc
Comment (adisuissa) [inline source/common/rds/rds_route_config_subscription.cc:103]: Seems not related to this PR, so flagging to make sure the change was intended.

File: test/common/router/scoped_config_impl_test.cc
Comment (adisuissa) [inline test/common/router/scoped_config_impl_test.cc:399]: nit: const

File: source/common/router/scoped_rds.cc
Comment (adisuissa) [reply source/common/router/scoped_rds.cc:288]: Suggestion: checking if this can be also fixed in the server that provides the config, to avoid sending non-modified data on the wire.

File: source/common/router/scoped_rds.cc
Comment (adisuissa) [inline source/common/router/scoped_rds.cc:444]: Nice catch!

**Review by adisuissa — COMMENTED** [review]
Thanks for fixing this!
LGTM modulo small nits.
/assign-from @envoyproxy/senior-maintainers

File: source/common/router/scoped_rds.cc
Comment (stevenzzzz) [reply source/common/router/scoped_rds.cc:288]: ACK.

File: source/common/router/scoped_rds.cc
Comment (stevenzzzz) [reply source/common/router/scoped_rds.cc:289]: hmm, we definitely saves the source config for cluster and RDS route tables. 

For RDS it's in ConfigInfo: 
```
  struct ConfigInfo {
    // A reference to the currently loaded route configuration. Do not hold this reference beyond
    // the caller of configInfo()'s scope.
    const Protobuf::Message& config_;

    // The discovery version that supplied this route. This will be set to "" in the case of
    // static clusters.
    const std::string version_;
  };

```

for CDS it's in ClusterData: 

```
  struct ClusterData : public ClusterManagerCluster {
    ClusterData(const envoy::config::cluster::v3::Cluster& cluster_config,
                const uint64_t cluster_config_hash, const std::string& version_info,
                bool added_via_api, ClusterSharedPtr&& cluster, TimeSource& time_source)
        : cluster_config_(cluster_config), config_hash_(cluster_config_hash),
          version_info_(version_info), added_via_api_(added_via_api), cluster_(std::move(cluster)),
          last_updated_(time_source.systemTime()) {}

    bool blockUpdate(uint64_t hash) { return !added_via_api_ || config_hash_ == hash; }

    const envoy::config::cluster::v3::Cluster cluster_config_;
```

File: test/common/router/scoped_config_impl_test.cc
Comment (stevenzzzz) [reply test/common/router/scoped_config_impl_test.cc]: Thanks!

File: test/common/router/scoped_config_impl_test.cc
Comment (stevenzzzz) [reply test/common/router/scoped_config_impl_test.cc]: done

**Review by stevenzzzz — COMMENTED** [review]
thanks!

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**
